src/node/miner.cpp

```diff
@@ -56,34 +56,22 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman)
     block.hashMerkleRoot = BlockMerkleRoot(block);
 }
 
-BlockAssembler::Options::Options()
-{
-    blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
-    nBlockMaxWeight = DEFAULT_BLOCK_MAX_WEIGHT;
-    test_block_validity = true;
-}
-
 BlockAssembler::BlockAssembler(Chainstate& chainstate, const CTxMemPool* mempool, const Options& options)
-    : test_block_validity{options.test_block_validity},
-      chainparams{chainstate.m_chainman.GetParams()},
+    : chainparams{chainstate.m_chainman.GetParams()},
       m_mempool(mempool),
-      m_chainstate(chainstate)
+      m_chainstate(chainstate),
+      m_options(options)
 {
-    blockMinFeeRate = options.blockMinFeeRate;
-    // Limit weight to between 4K and MAX_BLOCK_WEIGHT-4K for sanity:
-    nBlockMaxWeight = std::max<size_t>(4000, std::min<size_t>(MAX_BLOCK_WEIGHT - 4000, options.nBlockMaxWeight));
+    // Limit weight to between 4K and DEFAULT_BLOCK_MAX_WEIGHT for sanity:
+    m_options.nBlockMaxWeight = std::clamp<size_t>(options.nBlockMaxWeight, 4000, DEFAULT_BLOCK_MAX_WEIGHT);
 }
 
 void ApplyArgsManOptions(const ArgsManager& args, BlockAssembler::Options& options)
 {
     // Block resource limits
-    // If -blockmaxweight is not given, limit to DEFAULT_BLOCK_MAX_WEIGHT
-    options.nBlockMaxWeight = args.GetIntArg("-blockmaxweight", DEFAULT_BLOCK_MAX_WEIGHT);
-    if (args.IsArgSet("-blockmintxfee")) {
-        std::optional<CAmount> parsed = ParseMoney(args.GetArg("-blockmintxfee", ""));
-        options.blockMinFeeRate = CFeeRate{parsed.value_or(DEFAULT_BLOCK_MIN_TX_FEE)};
-    } else {
-        options.blockMinFeeRate = CFeeRate{DEFAULT_BLOCK_MIN_TX_FEE};
+    options.nBlockMaxWeight = args.GetIntArg("-blockmaxweight", options.nBlockMaxWeight);
```

Comment:
This (and below) is a potential behaviour change: if you call `ApplyArgsManOptions`, then change the arguments, and call `ApplyArgsManOptions` again on the same `options` object it won't reset values back to the defaults any more. Probably no big deal, but not really a refactor.

Reply:
True that there is a difference. I think this is better as it matches other `ApplyArgsManOptions` around the codebase - convention seems to be to default to the value passed in and not look at default constants. For example
https://github.com/bitcoin/bitcoin/blob/7386da7a0b08cd2df8ba88dae1fab9d36424b15c/src/node/mempool_args.cpp#L29-L38
https://github.com/bitcoin/bitcoin/blob/7386da7a0b08cd2df8ba88dae1fab9d36424b15c/src/node/validation_cache_args.cpp#L20-L32

Reply:
That's a good observation, I hadn't considered that, thanks @ajtowns. I agree with @glozow that I'd prefer this behaviour (I intentionally brought it in line with existing `ApplyArgsManOptions()` functions and `Options` objects). In practice, there is no actual behaviour change in the current code as we never use `node::ApplyArgsManOptions()` in the way you describe here. However, I'm not quite sure if it would be inappropriate to label this commit and PR with `refactor: ` - what do you think?

Reply:
Doesn't it make more sense for `ApplyArgsManOptions(args, foo)` to consistently reset `foo` to the same thing you'd get if you'd just created `foo` as a new object? TBH, I think just having an `explicit Options(const ArgsManager& args);` constructor and writing `Options cache_sizes{gArgs}` would probably be a better api...

Reply:
>  TBH, I think just having an explicit Options(const ArgsManager& args); constructor and writing Options cache_sizes{gArgs} would probably be a better api...

I'm mostly going with the idea from this commit message (https://github.com/bitcoin/bitcoin/commit/f1941e8bfd2eecc478c7660434b1ebf6a64095a0):

> With this patch, most CTxMemPool construction are along the lines of:
>
>    MemPoolOptions mempool_opts{...default overrides...};
>    ApplyArgsManOptions(argsman, mempool_opts);
>    ...hard overrides...
>    CTxMemPool pool{mempool_opts};
>
> This "compositional" style of building options means that we can omit
> unnecessary/irrelevant steps wherever we want but also maintain full
> customizability.
>
> For example:
>
> - For users of libbitcoinkernel, where we eventually want to remove
>  ArgsManager, they simply won't call (or even know about)
>  ApplyArgsManOptions.

As in, {mempool, block assembler, cache, etc.} `::Options` struct living in kernel/ shouldn't know about the `ArgsManager` which lives in node/. Otherwise I agree a ctor using argsman makes sense.